### PR TITLE
Enable _GLIBCXX_ASSERTIONS and fix unittest problems

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

**Comment by athanatos** [comment]
We actually enable _GLIBCXX_ASSERTIONS in centos8 packages already, the build change here simply enables it generally.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -108,6 +108,12 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
   endif()
 endif(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
 
+CHECK_C_COMPILER_FLAG("-D_GLIBCXX_ASSERTIONS" HAS_GLIBC_DEBUG)
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i don't think this check makes sense. as `HAS_GLIBC_DEBUG` does not imply that `_GLIBCXX_ASSERTIONS` is supported by the C++ standard library being used.

instead, i think we could just add `_LIBCPP_DEBUG` for clang and add `_GLIBCXX_DEBUG` for GCC if `CMAKE_BUILD_TYPE` is `Debug`.

as `_GLIBCXX_ASSERTIONS` kicks in even when a smart point is dereferenced, i'd suggest avoid defining it in release builds.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -108,6 +108,12 @@ if(CMAKE_CXX_COMPILER_ID STREQUAL GNU)
...
+if(HAS_GLIBC_DEBUG)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_GLIBCXX_ASSERTIONS")
```
Comment (tchaikov) [inline src/CMakeLists.txt]: i don't think `-D_GLIBCXX_ASSERTIONS` helps when compiling C code.

Comment (athanatos) [reply src/CMakeLists.txt]: True.

File: src/CMakeLists.txt (hunk 1 above)
Comment (athanatos) [reply src/CMakeLists.txt]: I think I chose the name HAS_GLIBC_DEBUG poorly.  I'm replacing it with HAS_GLIBCXX_ASSERTIONS.  As I indicated above, we're actually already defining it for release rpm builds.  I'm actually fine with restricting it to debug builds, however, if that's what we want to do.

Comment (tchaikov) [reply src/CMakeLists.txt]: yes, can we just restrict it to Debug builds?

**Comment by athanatos** [comment]
retest this please

**Pull request merged**